pkg/server/registry/apigroups/acorn/images/copy.go

```diff
@@ -0,0 +1,353 @@
+			}
+		}
+	}()
+
+	go func() {
+		defer func() {
+			wg.Wait()
+			close(progress2)
+		}()
+
+		// Don't write error to chan because it already gets sent to the progress chan by remote.WriteIndex()
+		if err := remote.WriteIndex(destRef, sourceIndex, destOpts...); err != nil {
+			handleWriteIndexError(err, progress)
+		}
+	}()
+
+	return typed.Every(500*time.Millisecond, progress2), nil
+}
+
+func (i *ImageCopy) RepoCopy(ctx context.Context, namespace, source, dest string, sourceAuth, destAuth *apiv1.RegistryAuth, force bool) (<-chan ImageProgress, error) {
+	sourceRepo, err := name.NewRepository(source)
+	if err != nil {
+		return nil, err
+	}
+
+	destRepo, err := name.NewRepository(dest)
+	if err != nil {
+		return nil, err
+	}
+
+	sourceOpts, err := images.GetAuthenticationRemoteOptionsWithLocalAuth(ctx, sourceRepo.Registry, sourceAuth, i.client, namespace, i.transportOpt)
+	if err != nil {
+		return nil, err
+	}
+
+	destOpts, err := images.GetAuthenticationRemoteOptionsWithLocalAuth(ctx, destRepo.Registry, destAuth, i.client, namespace, i.transportOpt)
+	if err != nil {
+		return nil, err
+	}
+
+	sourceTags, err := remote.List(sourceRepo, sourceOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	sourceIndexes := make([]ggcrv1.ImageIndex, len(sourceTags))
+	for i, tag := range sourceTags {
+		sourceIndexes[i], err = remote.Index(sourceRepo.Tag(tag), sourceOpts...)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Don't copy tags that already exist in the destination if force is not set
+	if !force {
+		var newIndexSlice []ggcrv1.ImageIndex
+		for i, imageIndex := range sourceIndexes {
+			if err := errIfImageExistsAndIsDifferent(imageIndex, destRepo.Tag(sourceTags[i]), destOpts); err == nil {
+				newIndexSlice = append(newIndexSlice, imageIndex)
+			}
+		}
+		sourceIndexes = newIndexSlice
+	}
+
+	// metachannel is used to send another channel with updates for each tag to be copied
+	metachannel := make(chan chan ggcrv1.Update)
+	// progress is the channel returned by this function and used to write websocket messages to the client
+	progress := make(chan ImageProgress)
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+
+	// We create an RWMutex for the currentTag variable, since it is used by both goroutines
+	m := sync.RWMutex{}
+	currentTag := ""
+
+	go func() {
+		defer wg.Done()
+		for currProgress := range metachannel {
+			for update := range currProgress {
+				var errString = ""
+				if update.Error != nil {
+					errString = update.Error.Error()
+				}
+				m.RLock()
+				progress <- ImageProgress{
+					Total:       update.Total,
+					Complete:    update.Complete,
+					Error:       errString,
+					CurrentTask: fmt.Sprintf("Copying %s:%s to %s:%s", source, currentTag, dest, currentTag),
+				}
+				m.RUnlock()
+			}
+		}
+	}()
+
+	go func() {
+		defer func() {
+			close(metachannel)
+			wg.Wait()
+			close(progress)
+		}()
+
+		for i, imageIndex := range sourceIndexes {
+			currProgress := make(chan ggcrv1.Update)
```

Comment:
How do you feel about adding a comment here that `remote.WithProgress` will close this channel? Otherwise, it looks like it leaks.